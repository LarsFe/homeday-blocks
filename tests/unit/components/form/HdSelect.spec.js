import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdSelect from '@/components/form/HdSelect.vue';
import ITEMS from '@/stories/mocks/FORM_ITEMS';
import FIELD_CLASSES from './FIELD_CLASSES';

const ERROR_SELECTOR = '.field__error';
const HELPER_SELECTOR = '.field__error--helper';

describe('HdSelect', () => {
  const wrapperFactory = wrapperFactoryBuilder(HdSelect, {
    propsData: {
      name: 'testSelect',
      label: 'test select label',
      options: ITEMS,
    },
  });

  let wrapper;
  beforeEach(() => {
    wrapper = wrapperFactory();
  });

  test('is rendered as expected', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('renders with preselected item', () => {
    const ITEM_VALUE = ITEMS[2].value;
    wrapper = wrapperFactory({
      propsData: {
        value: ITEM_VALUE,
      },
    });
    expect(wrapper.find('select').element.value).toBe(ITEM_VALUE);
  });

  test('can change the selected item', () => {
    const ITEM_VALUE = ITEMS[2].value;
    wrapper.setProps({ value: ITEM_VALUE });
    expect(wrapper.find('select').element.value).toBe(ITEM_VALUE);
  });

  test('emits `input` event with the right payload', () => {
    const ITEM_INDEX = 1;
    wrapper.findAll('option').at(ITEM_INDEX).setSelected();
    expect(wrapper.emitted('input')[0][0]).toEqual(ITEMS[ITEM_INDEX].value);
  });

  test('allows setting a helper', () => {
    const TEST_HTML = '<b>test</b> test';
    wrapper.vm.showHelper(TEST_HTML);
    expect(wrapper.find(HELPER_SELECTOR).element.innerHTML).toEqual(TEST_HTML);
  });

  test('validates requiredness', () => {
    wrapper.setProps({
      value: undefined,
      required: true,
    });
    expect(wrapper.vm.validate()).toBe(false);
    expect(wrapper.find(ERROR_SELECTOR).text()).toBeTruthy();
    expect(wrapper.classes()).toContain(FIELD_CLASSES.INVALID);
    wrapper.setProps({
      required: false,
    });
    expect(wrapper.vm.validate()).toBe(true);
  });

  test('Supports disabling', () => {
    wrapper.setProps({
      disabled: true,
    });
    expect(wrapper.find('select').attributes().disabled).toBe('disabled');
  });
});
