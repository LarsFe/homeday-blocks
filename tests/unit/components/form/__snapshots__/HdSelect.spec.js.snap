// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdSelect is rendered as expected 1`] = `
<div class="field field--select"><select id="testSelect" name="testSelect" class="field__input">
    <option value="jon">Jon "Snow"</option>
    <option value="daenerys">Daenerys</option>
    <option value="nightKing">The Night King</option>
    <option value="arya">Arya</option>
    <option value="cersei">Cersei</option>
  </select> <label for="testSelect" class="field__label">test select label</label>
  <!----> <span class="field__border"></span></div>
`;
